src/Sylius/Component/Core/Resolver/CategoryBasedDefaultShippingMethodResolver.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Resolver;
+
+use Sylius\Component\Addressing\Matcher\ZoneMatcherInterface;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Core\Model\AddressInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\ShipmentInterface as CoreShipmentInterface;
+use Sylius\Component\Core\Repository\ShippingMethodRepositoryInterface;
+use Sylius\Component\Shipping\Checker\ShippingMethodEligibilityCheckerInterface;
+use Sylius\Component\Shipping\Exception\UnresolvedDefaultShippingMethodException;
+use Sylius\Component\Shipping\Model\ShipmentInterface;
+use Sylius\Component\Shipping\Model\ShippingMethodInterface;
+use Sylius\Component\Shipping\Resolver\DefaultShippingMethodResolverInterface;
+use Webmozart\Assert\Assert;
+
+final class CategoryBasedDefaultShippingMethodResolver implements DefaultShippingMethodResolverInterface
+{
+    /**
+     * @var ShippingMethodRepositoryInterface
+     */
+    private $shippingMethodRepository;
+
+    /**
+     * @var ShippingMethodEligibilityCheckerInterface
+     */
+    private $shippingMethodEligibilityChecker;
+
+    /**
+     * @var ZoneMatcherInterface
+     */
+    private $zoneMatcher;
+
+    /**
+     * @param ShippingMethodRepositoryInterface $shippingMethodRepository
+     * @param ShippingMethodEligibilityCheckerInterface $shippingMethodEligibilityChecker
+     * @param ZoneMatcherInterface $zoneMatcher
+     */
+    public function __construct(
+        ShippingMethodRepositoryInterface $shippingMethodRepository,
+        ShippingMethodEligibilityCheckerInterface $shippingMethodEligibilityChecker,
+        ?ZoneMatcherInterface $zoneMatcher
```

Comment:
Should it be an optional parameter? Either we need to remove question mark and maybe change the name to `CategoryAndZoneBasedDefaultShippingMethodResolver` or change `ZoneMatcherInterface` typehint to `ZoneMatcherInterface|null` and support nullable value in line 97.

Reply:
And change the name to `CategoryAndPerhapsZoneBasedDefaultShippingMethodResolver` 🐋

Reply:
It should not be optional. Will be fixed in a minute

Reply:
`EligibleDefaultShippingMethodResolver`? And yeah, zone matcher should be required.